<a href="https://colab.research.google.com/github/yohanesnuwara/reservoir-engineering/blob/master/Reservoir%20Simulation%20Ertekin/Unit%205%20Finite-Difference%20Approximation%20to%20Linear-Flow%20Equations/fd1d_linearflow_explicit.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Unit 5 Finite-Difference Approximations to Linear-Flow Problems**

## Explicit Formulation (*Forward-Difference Approximation*)

![Fig 5 29 Explicit Method](https://user-images.githubusercontent.com/51282928/76113357-94c97880-6016-11ea-8a3f-7658d9208cf9.PNG)

$$p_{i}^{n+1} = p^i_n + (\frac{\alpha_c B_l \Delta t}{V_b \phi c_l})_i \cdot q_{lsc_i} + (\frac{\alpha_c B_l \Delta t}{V_b \phi c_l})_i \cdot [T_{lx_{i+1/2}}^{n} p_{i+1}^{n}-(T_{lx_{i+1/2}}^{n} + T_{lx_{i-1/2}}^{n})p_{i}^{n}+T_{lx_{i-1/2}}p_{i-1}^{n}]$$

In the following codes, $(\frac{\alpha_c B_l \Delta t}{V_b \phi c_l})_i$ will be written as `factor`

Transmissibility of coupling cells, $T_{lx_{i\pm1/2}}^{n}$, written as `T_min_half` or `T_plus_half`

In [0]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

In [2]:
!git clone https://github.com/yohanesnuwara/reservoir-engineering

fatal: destination path 'reservoir-engineering' already exists and is not an empty directory.


## Example 5.8 Five reservoir gridblocks with no flow at both boundaries

![Example 5 8](https://user-images.githubusercontent.com/51282928/75949840-09e36380-5eda-11ea-824c-dd5c9ac6f6d5.PNG)

### For timestep 10 days

In [56]:
"Task. determine pressure distribution during first year of production with timestep 10 days"

# known
p_initial = 6000 # initial pressure, in psia
delta_x = 1000 # ft
delta_y = 1000
delta_z = 75
ngrid = 5
grid_loc = 4 # grid location where production well is located
B = 1 # phase FVF, assumed constant over pressure, rb/stb
c = 3.5E-06 # phase compressibility, psi^-1
k_x = 15 # perm in x direction, md
poro = 0.18
mu = 10 # phase viscosity, cp

delta_t = 10 # days
qsc = -150 # minus means production, stb/d

# conversion
k_x = 15 * 1E-03 # darcy to mD

# calculate factor
Vb = delta_x * delta_y * delta_z
alpha = 5.615 # volume conversion factor, is a constant
factor = (alpha * B * delta_t) / (Vb * poro * c)
factor

1.1883597883597885

Because of uniform gridblocks, the equation for `tr_plus_coupling` and `tr_min_coupling` become:

$$T_{lx_{i+1/2}}^{n} = T_{lx_{i-1/2}}^{n} = (\beta_c \frac{A_x k_x}{\mu_l B_l \Delta x})_{i+1/2}^{n} = (\beta_c \frac{A_x k_x}{\mu_l B_l \Delta x})_{i-1/2}^{n}$$

In [57]:
# calculate transmissibility of coupling cells
beta = 1.127 # transmissibility conversion factor, is a constant
Ax = delta_y * delta_z
T_plus_half = beta * ((Ax * k_x) / (mu * B * delta_x))
T_min_half = T_plus_half
T_min_half, T_plus_half

(0.1267875, 0.1267875)

There are 5 grids (`grid_1`, `grid_2`, `grid_3`, `grid_4`, `grid_5`), so the transmissibility coefficients will be: $$T_{lx_{i+1/2}}^{n}=(T_{lx_{1+1/2}}^{n}, T_{lx_{2+1/2}}^{n}, T_{lx_{3+1/2}}^{n}, T_{lx_{4+1/2}}^{n}, T_{lx_{5+1/2}}^{n})$$

and

$$T_{lx_{i-1/2}}^{n}=(T_{lx_{1/2}}^{n}, T_{lx_{1+1/2}}^{n}, T_{lx_{2+1/2}}^{n}, T_{lx_{3+1/2}}^{n}, T_{lx_{4+1/2}}^{n})$$

The values of $T_{lx_{5+1/2}}^{n}=0$ and $T_{lx_{1/2}}^{n}=0$ (the edge of gridblocks), so the values of each are:

$$T_{lx_{i+1/2}}^{n}=(0.127, 0.127, 0.127, 0.127, 0)$$

and

$$T_{lx_{i-1/2}}^{n}=(0, 0.127, 0.127, 0.127, 0.127)$$




In [58]:
q = np.full(ngrid-1, T_min_half)
Ti_plus_halves = np.append(q, [0])
print(Ti_plus_halves)
p = np.full(ngrid-1, T_plus_half)
Ti_min_halves = np.append([0], p)
print(Ti_min_halves)
print("At grid 1, the coupling transmissibility coeffs are:", Ti_min_halves[0], "for T_min_half and:", Ti_plus_halves[0], "for T_plus_half.")
print("At grid 3, the coupling transmissibility coeffs are:", Ti_min_halves[2], "for T_min_half and:", Ti_plus_halves[2], "for T_plus_half.")
print("At grid 5, the coupling transmissibility coeffs are:", Ti_min_halves[4], "for T_min_half and:", Ti_plus_halves[4], "for T_plus_half.")

[0.1267875 0.1267875 0.1267875 0.1267875 0.       ]
[0.        0.1267875 0.1267875 0.1267875 0.1267875]
At grid 1, the coupling transmissibility coeffs are: 0.0 for T_min_half and: 0.1267875 for T_plus_half.
At grid 3, the coupling transmissibility coeffs are: 0.1267875 for T_min_half and: 0.1267875 for T_plus_half.
At grid 5, the coupling transmissibility coeffs are: 0.1267875 for T_min_half and: 0.0 for T_plus_half.


Requires array for $q_{sc}$

In [0]:
qsc = [0, 0, 0, -150, 0] # production well in grid 4

Calculate $p_{i}^{n+1}$ for each grid in each time

In [60]:
pi = np.full(ngrid, p_initial) # array of pressure in each grid [6000, 6000, 6000, 6000, 6000]
time = np.arange(15, 370, delta_t)

pi_arr = []
min_arr = []
plus_arr = []

for j in range(len(time)):
  pnew_arr = []

  minus = pi[0]
  plus = pi[-1]
  minus_arr = []
  plusus_arr = []
  for i, obj in enumerate(pi):
    if i > 0:
      minus = pi[i-1]
    if i < (len(pi) - 1):
      plus = pi[i+1]

    pnew = pi[i] + (factor * ((Ti_plus_halves[i] * plus) - ((Ti_plus_halves[i] + Ti_min_halves[i]) * pi[i]) + (Ti_min_halves[i] * minus))) + (factor * qsc[i])
    pnew_arr.append(float(pnew))

    minus_arr.append(float(minus))
    plusus_arr.append(float(plus))

  pi = pnew_arr

  min_arr.append(minus_arr)
  plus_arr.append(plusus_arr)
  pi_arr.append(pi)

df = pd.DataFrame.from_records(pi_arr)
df = pd.DataFrame(pd.np.column_stack([time, df]), columns=['time', 'grid 1', 'grid 2', 'grid 3', 'grid 4', 'grid 5'])
df

time       grid 1       grid 2       grid 3       grid 4       grid 5
0    15.0  6000.000000  6000.000000  6000.000000  5821.746032  6000.000000
1    25.0  6000.000000  6000.000000  5973.142623  5697.206817  5973.142623
2    35.0  6000.000000  5995.953421  5935.614184  5602.102885  5931.567605
3    45.0  5999.390305  5987.471853  5894.455577  5523.738961  5881.927430
4    55.0  5997.594562  5975.252912  5852.614698  5455.308514  5827.959472
5    65.0  5994.228364  5960.141312  5811.230704  5393.063347  5771.812463
6    75.0  5989.092497  5942.840943  5770.662014  5334.880120  5714.746649
7    85.0  5982.123813  5923.867570  5730.945175  5279.519220  5657.512476
8    95.0  5973.346394  5903.577533  5691.996659  5226.233153  5600.560547
9   105.0  5962.834378  5882.210835  5653.699174  5174.554980  5544.160950
10  115.0  5950.686896  5859.928656  5615.936579  5124.181492  5488.472727
11  125.0  5937.012427  5836.841041  5578.606333  5074.907310  5433.585270
12  135.0  5921.919688  5813.025772  5541.622429  5026.586963  5379.543561
13  145.0  5905.512733  5788.540612  5504.914580  4979.112636  5326.363884
14  155.0  5887.888641  5763.431006  5468.426137  4932.400864  5274.043828
15  165.0  5869.136713  5737.734796  5432.111794  4886.384438  5222.568767
16  175.0  5849.338495  5711.485050  5395.935471  4841.007368  5171.916155
17  185.0  5828.568232  5684.711722  5359.868508  4796.221706  5122.058404
18  195.0  5806.893491  5657.442606  5323.888171  4751.985475  5072.964860
19  205.0  5784.375851  5629.703878  5287.976437  4708.261306  5024.603163
20  215.0  5761.071554  5601.520386  5252.119030  4665.015497  4976.940199
21  225.0  5737.032112  5572.915816  5216.304641  4622.217364  4929.942765
22  235.0  5712.304847  5543.912773  5180.524315  4579.838760  4883.578035
23  245.0  5686.933353  5514.532830  5144.770959  4537.853727  4837.813892
24  255.0  5660.957910  5484.796561  5109.038959  4496.238221  4792.619144
25  265.0  5634.415826  5454.723560  5073.323867  4454.969896  4747.963677
26  275.0  5607.341742  5424.332470  5037.622163  4414.027934  4703.818548
27  285.0  5579.767888  5393.641005  5001.931060  4373.392899  4660.156038
28  295.0  5551.724305  5362.665976  4966.248350  4333.046615  4616.949675
29  305.0  5523.239044  5331.423324  4930.572286  4292.972062  4574.174237
30  315.0  5494.338330  5299.928147  4894.901483  4253.153285  4531.805740
31  325.0  5465.046709  5268.194737  4859.234847  4213.575316  4489.821407
32  335.0  5435.387187  5236.236614  4823.571509  4174.224100  4448.199638
33  345.0  5405.381336  5204.066557  4787.910784  4135.086431  4406.919972
34  355.0  5375.049406  5171.696644  4752.252126  4096.149897  4365.963039
35  365.0  5344.410415  5139.138279  4716.595106  4057.402826  4325.310518

### For timestep 15 days

In [61]:
delta_t = 15 # days

# calculate factor
Vb = delta_x * delta_y * delta_z
alpha = 5.615 # volume conversion factor, is a constant
factor = (alpha * B * delta_t) / (Vb * poro * c)

pi = np.full(ngrid, 6000) # array of pressure in each grid [6000, 6000, 6000, 6000, 6000]
time = np.arange(15, 370, delta_t)

pi_arr = []
min_arr = []
plus_arr = []

for j in range(len(time)):
  pnew_arr = []

  minus = pi[0]
  plus = pi[-1]
  minus_arr = []
  plusus_arr = []
  for i, obj in enumerate(pi):
    if i > 0:
      minus = pi[i-1]
    if i < (len(pi) - 1):
      plus = pi[i+1]

    pnew = pi[i] + (factor * ((Ti_plus_halves[i] * plus) - ((Ti_plus_halves[i] + Ti_min_halves[i]) * pi[i]) + (Ti_min_halves[i] * minus))) + (factor * qsc[i])
    pnew_arr.append(float(pnew))

    minus_arr.append(float(minus))
    plusus_arr.append(float(plus))

  pi = pnew_arr

  min_arr.append(minus_arr)
  plus_arr.append(plusus_arr)
  pi_arr.append(pi)

df = pd.DataFrame.from_records(pi_arr)
df = pd.DataFrame(pd.np.column_stack([time, df]), columns=['time', 'grid 1', 'grid 2', 'grid 3', 'grid 4', 'grid 5'])
df

time       grid 1       grid 2       grid 3       grid 4       grid 5
0    15.0  6000.000000  6000.000000  6000.000000  5732.619048  6000.000000
1    30.0  6000.000000  6000.000000  5939.570902  5586.096291  5939.570902
2    45.0  6000.000000  5986.342797  5873.341517  5478.488514  5859.684314
3    60.0  5996.913421  5963.890663  5809.641971  5386.497501  5773.532634
4    75.0  5989.450154  5936.493147  5748.870517  5302.220177  5686.061243
5    90.0  5977.481672  5906.058211  5690.329283  5222.533397  5599.311723
6   105.0  5961.339702  5873.444635  5633.361205  5146.029383  5514.158408
7   120.0  5941.475087  5839.049494  5577.482141  5071.985790  5430.959868
8   135.0  5918.326519  5803.082860  5522.353273  4999.978397  5349.830380
9   150.0  5892.281020  5765.682420  5467.740531  4929.723985  5270.762520
10  165.0  5863.669261  5726.958194  5413.482758  4861.012778  5193.686532
11  180.0  5832.772047  5687.008784  5359.469095  4793.677629  5118.501016
12  195.0  5799.829003  5645.926630  5305.623300  4727.578973  5045.089713
13  210.0  5765.046490  5603.799315  5251.892943  4662.596824  4973.331095
14  225.0  5728.604024  5560.709621  5198.241970  4598.626115  4903.103984
15  240.0  5690.659259  5516.735337  5144.645587  4535.573735  4834.290844
16  255.0  5651.351801  5471.949117  5091.086743  4473.356492  4766.779657
17  270.0  5610.806121  5426.418471  5037.553715  4411.899628  4700.464921
18  285.0  5569.133821  5380.205878  4984.038438  4351.135684  4635.248083
19  300.0  5526.435398  5333.368975  4930.535369  4291.003595  4571.037615
20  315.0  5482.801662  5285.960805  4877.040696  4231.447961  4507.748877
21  330.0  5438.314890  5238.030099  4823.551795  4172.418431  4445.303834
22  345.0  5393.049776  5189.621562  4770.066864  4113.869185  4383.630709
23  360.0  5347.074237  5140.776166  4716.584663  4055.758485  4322.663593

## Example 5.9 Similar to Ex 5.8, but **constant pressure at left boundary of reservoir**

![Example 5 9](https://user-images.githubusercontent.com/51282928/76109912-081bbc00-6010-11ea-9f00-023e9d954ac6.PNG)

The only modification is in the line:

```
pnew = pi[i] + (factor * ((Ti_plus_halves[i] * plus) - ((Ti_plus_halves[i] + Ti_min_halves[i]) * pi[i]) + (Ti_min_halves[i] * minus))) + (factor * qsc[i])
pnew_arr.append(float(pnew))
pnew_arr[0] = pi[0]
```
Where `pnew_arr[0] = pi[0]` means the pressure in the first gridblock is kept constant, equals to initial pressure `pi` 6000 psi.


### For timestep 15 days

In [62]:
pi = np.full(ngrid, p_initial) # array of pressure in each grid [6000, 6000, 6000, 6000, 6000]
time = np.arange(15, 370, delta_t)

pi_arr = []
min_arr = []
plus_arr = []

for j in range(len(time)):
  pnew_arr = []

  plus = pi[-1]
  minus_arr = []
  plusus_arr = []
  for i, obj in enumerate(pi):
    if i > 0:
      minus = pi[i-1]
    if i < (len(pi) - 1):
      plus = pi[i+1]

    pnew = pi[i] + (factor * ((Ti_plus_halves[i] * plus) - ((Ti_plus_halves[i] + Ti_min_halves[i]) * pi[i]) + (Ti_min_halves[i] * minus))) + (factor * qsc[i])
    pnew_arr.append(float(pnew))
    pnew_arr[0] = pi[0]

    minus_arr.append(float(minus))
    plusus_arr.append(float(plus))

  pi = pnew_arr

  min_arr.append(minus_arr)
  plus_arr.append(plusus_arr)
  pi_arr.append(pi)

df = pd.DataFrame.from_records(pi_arr)
df = pd.DataFrame(pd.np.column_stack([time, df]), columns=['time', 'grid 1', 'grid 2', 'grid 3', 'grid 4', 'grid 5'])
df

time  grid 1       grid 2       grid 3       grid 4       grid 5
0    15.0  6000.0  6000.000000  6000.000000  5732.619048  6000.000000
1    30.0  6000.0  6000.000000  5939.570902  5586.096291  5939.570902
2    45.0  6000.0  5986.342797  5873.341517  5478.488514  5859.684314
3    60.0  6000.0  5963.890663  5809.641971  5386.497501  5773.532634
4    75.0  6000.0  5937.190726  5748.870517  5302.220177  5686.061243
5    90.0  6000.0  5908.824784  5690.486939  5222.533397  5599.311723
6   105.0  6000.0  5880.085553  5634.072855  5146.065014  5514.158408
7   120.0  6000.0  5851.586875  5579.381045  5072.166151  5430.967921
8   135.0  6000.0  5823.609260  5526.268116  5000.508212  5349.877375
9   150.0  6000.0  5796.274015  5474.644619  4930.909711  5270.918634
10  165.0  6000.0  5769.627402  5424.447941  4863.258178  5194.075343
11  180.0  6000.0  5743.680620  5375.628803  4797.474136  5119.309423
12  195.0  6000.0  5718.428670  5328.144771  4733.494288  5046.573441
13  210.0  6000.0  5693.859222  5281.957157  4671.263638  4975.816379
14  225.0  6000.0  5669.956774  5237.029543  4610.731772  4906.986317
15  240.0  6000.0  5646.704603  5193.327076  4551.851103  4840.031679
16  255.0  6000.0  5624.085666  5150.816122  4494.576017  4774.901788
17  270.0  6000.0  5602.083024  5109.464089  4438.862464  4711.547112
18  285.0  6000.0  5580.680026  5069.239334  4384.667747  4649.919359
19  300.0  6000.0  5559.860398  5030.111102  4331.950399  4589.971500
20  315.0  6000.0  5539.608271  4992.049493  4280.670120  4531.657764
21  330.0  6000.0  5519.908191  4955.025424  4230.787722  4474.933615
22  345.0  6000.0  5500.745117  4919.010611  4182.265094  4419.755728
23  360.0  6000.0  5482.104413  4883.977541  4135.065165  4366.081954

## Example 5.10 Halving the grid spacing of Ex 5.9 (more refined grids)

![Example 5 10](https://user-images.githubusercontent.com/51282928/76109917-0c47d980-6010-11ea-95bf-87492c658368.PNG)

### For timestep 15 days

In [63]:
# known
p_initial = 6000 # initial pressure, in psia
delta_x = 500 # ft, half the previous one
delta_y = 1000
delta_z = 75
ngrid = 10 # now twice the number of previous gridblocks
grid_loc = 4 # grid location where production well is located
B = 1 # phase FVF, assumed constant over pressure, rb/stb
c = 3.5E-06 # phase compressibility, psi^-1
k_x = 15 # perm in x direction, md
poro = 0.18
mu = 10 # phase viscosity, cp

delta_t = 15 # days
qsc = -150 # minus means production, stb/d

# conversion
k_x = 15 * 1E-03 # darcy to mD

# calculate factor
Vb = delta_x * delta_y * delta_z
alpha = 5.615 # volume conversion factor, is a constant
factor = (alpha * B * delta_t) / (Vb * poro * c)
factor

3.5650793650793653

In [64]:
# calculate transmissibility of coupling cells
beta = 1.127 # transmissibility conversion factor, is a constant
Ax = delta_y * delta_z
T_plus_half = beta * ((Ax * k_x) / (mu * B * delta_x))
T_min_half = T_plus_half
T_min_half, T_plus_half

(0.253575, 0.253575)

In [67]:
q = np.full(ngrid-1, T_min_half)
Ti_plus_halves = np.append(q, [0])
print(Ti_plus_halves)
p = np.full(ngrid-1, T_plus_half)
Ti_min_halves = np.append([0], p)
print(Ti_min_halves)
print("At grid 1, the coupling transmissibility coeffs are:", Ti_min_halves[0], "for T_min_half and:", Ti_plus_halves[0], "for T_plus_half.")
print("At grid 3, the coupling transmissibility coeffs are:", Ti_min_halves[2], "for T_min_half and:", Ti_plus_halves[2], "for T_plus_half.")
print("At grid 5, the coupling transmissibility coeffs are:", Ti_min_halves[4], "for T_min_half and:", Ti_plus_halves[4], "for T_plus_half.")

[0.253575 0.253575 0.253575 0.253575 0.253575 0.253575 0.253575 0.253575
 0.253575 0.      ]
[0.       0.253575 0.253575 0.253575 0.253575 0.253575 0.253575 0.253575
 0.253575 0.253575]
At grid 1, the coupling transmissibility coeffs are: 0.0 for T_min_half and: 0.253575 for T_plus_half.
At grid 3, the coupling transmissibility coeffs are: 0.253575 for T_min_half and: 0.253575 for T_plus_half.
At grid 5, the coupling transmissibility coeffs are: 0.253575 for T_min_half and: 0.253575 for T_plus_half.


In [0]:
qsc = [0, 0, 0, 0, 0, 0, -75, -75, 0, 0] # production well in grid 7 and 8

In [74]:
pi = np.full(ngrid, p_initial) # array of pressure in each grid [6000, 6000, 6000, 6000, 6000]
time = np.arange(15, 370, delta_t)

pi_arr = []
min_arr = []
plus_arr = []

for j in range(len(time)):
  pnew_arr = []

  plus = pi[-1]
  minus_arr = []
  plusus_arr = []
  for i, obj in enumerate(pi):
    if i > 0:
      minus = pi[i-1]
    if i < (len(pi) - 1):
      plus = pi[i+1]

    pnew = pi[i] + (factor * ((Ti_plus_halves[i] * plus) - ((Ti_plus_halves[i] + Ti_min_halves[i]) * pi[i]) + (Ti_min_halves[i] * minus))) + (factor * qsc[i])
    pnew_arr.append(float(pnew))
    pnew_arr[0] = pi[0]

    minus_arr.append(float(minus))
    plusus_arr.append(float(plus))

  pi = pnew_arr

  min_arr.append(minus_arr)
  plus_arr.append(plusus_arr)
  pi_arr.append(pi)

df = pd.DataFrame.from_records(pi_arr)
df = pd.DataFrame(pd.np.column_stack([time, df]), columns=['time', 'grid 1', 'grid 2', 'grid 3', 'grid 4', 'grid 5', 'grid 6', 'grid 7', 'grid 8', 'grid 9', 'grid 10'])
df

time  grid 1        grid 2  ...        grid 8        grid 9       grid 10
0    15.0  6000.0  6.000000e+03  ...  5.732619e+03  6.000000e+03  6.000000e+03
1    30.0  6000.0  6.000000e+03  ...  5.706954e+03  5.758284e+03  6.000000e+03
2    45.0  6000.0  6.000000e+03  ...  5.485976e+03  5.930397e+03  5.781485e+03
3    60.0  6000.0  6.000000e+03  ...  5.620358e+03  5.394015e+03  5.916103e+03
4    75.0  6000.0  6.000000e+03  ...  5.148360e+03  6.070608e+03  5.444128e+03
5    90.0  6000.0  5.838561e+03  ...  5.714705e+03  4.670534e+03  6.010475e+03
6   105.0  6000.0  6.475303e+03  ...  4.649321e+03  6.825807e+03  4.799148e+03
7   120.0  6000.0  4.295662e+03  ...  5.787900e+03  3.026101e+03  6.631278e+03
8   135.0  6000.0  1.067942e+04  ...  5.178875e+03  8.781943e+03  3.372144e+03
9   150.0  6000.0 -6.970620e+03  ...  1.386038e+03  6.341766e+02  8.262683e+03
10  165.0  6000.0  3.967533e+04  ...  2.094785e+04  8.210155e+03  1.366399e+03
11  180.0  6000.0 -8.054295e+04  ... -4.956521e+04  1.353836e+04  7.553257e+03
12  195.0  6000.0  2.238373e+05  ...  1.711932e+05 -4.891884e+04  1.296388e+04
13  210.0  6000.0 -5.383213e+05  ... -4.756604e+05  2.060086e+05 -4.297902e+04
14  225.0  6000.0  1.356126e+06  ...  1.344235e+06 -6.353190e+05  1.821095e+05
15  240.0  6000.0 -3.331704e+06  ... -3.645447e+06  1.893195e+06 -5.568581e+05
16  255.0  6000.0  8.238224e+06  ...  9.790637e+06 -5.328705e+06  1.658027e+06
17  270.0  6000.0 -2.028183e+07  ... -2.592411e+07  1.465552e+07 -4.658084e+06
18  285.0  6000.0  4.999876e+07  ...  6.809217e+07 -3.948886e+07  1.280170e+07
19  300.0  6000.0 -1.232486e+08  ... -1.775583e+08  1.050375e+08 -3.446975e+07
20  315.0  6000.0  3.041482e+08  ...  4.605307e+08 -2.765500e+08  9.164686e+07
21  330.0  6000.0 -7.513310e+08  ... -1.189155e+09  7.226375e+08 -2.412086e+08
22  345.0  6000.0  1.858385e+09  ...  3.059561e+09 -1.876983e+09  6.301227e+08
23  360.0  6000.0 -4.602587e+09  ... -7.848368e+09  4.852188e+09 -1.636339e+09

[24 rows x 11 columns]